2021q4/vdso.adoc

```diff
@@ -0,0 +1,79 @@
+== VDSO on amd64
+
+Contact: Konstantin Belousov <kib@FreeBSD.org>  
+
+VDSO AKA Virtual Dynamic Shared Object is a shared object (in the
+common speak, dynamic library) that is injected into the executed
+image by common efforts from kernel and dynamic linker.  It does not
+exists on disk as a standalone .so, and there are no instructions in
+the ELF format that cause the injection.  It is done by system to
+implement some functionality for the consumers.
+
+FreeBSD already has a lot of features that otherwise typically done
+using VDSO (in Linux), but really not requiring that complication.
+Main reason why it is possible is the often mentioned co-evolution of
+the kernel and C runtime.  We can naturally introduce features that
+require implementation both in kernel, and support in the userspace
+parts, since FreeBSD is developed as the whole.  Surprisingly, it also
+allows for kernel and dynamic linker to know much less (and not
+enforce anything) about userspace consumers of interfaces.
+
+For instance, syscall-less wall clock was implemented long time ago,
+by kernel providing time hands blob in the shared page, and C library
+knowing about the blob location and supported algorithms.  There is no
+need for VDSO that interposes some libc symbols, or provides services
+that are named by known symbols to userspace.
+
+From the all years of experience with this pseudo-VDSO approach, the
+only feature that was arguably impossible to implement without
+providing real VDSO, was the signal trampoline dwarf annotations, for
+the benefit of stack unwinders.
+
+When kernel delivers signal to userspace, it changes some key
+registers (like instruction pointer, stack pointer, whatever is needed
+by architecture), and pushes the saved image of whole usermode CPU
+state (context) into user stack.  Then, the control is passed to the
+small piece of code located on the shared page (signal trampoline),
+which calls into user function-handler, and on return from the handler
+issues sigreturn(2) syscall to reload old context.
+
+From this description, it is clear that the state of the machine
```

Comment:
Do you mean "stack state" here? Or is there more than that?

Reply:
I mean the architectural state of the CPU